mailpoet/lib/Automation/Integrations/WooCommerce/Fields/CustomerOrderFieldsFactory.php

```diff
@@ -0,0 +1,217 @@
+        Field::TYPE_DATETIME,
+        __('First paid order date', 'mailpoet'),
+        function (CustomerPayload $payload) {
+          $customer = $payload->getCustomer();
+          if (!$customer) {
+            return null;
+          }
+          return $this->getPaidOrderDate($customer, true);
+        }
+      ),
+      new Field(
+        'woocommerce:customer:last-paid-order-date',
+        Field::TYPE_DATETIME,
+        __('Last paid order date', 'mailpoet'),
+        function (CustomerPayload $payload) {
+          $customer = $payload->getCustomer();
+          if (!$customer) {
+            return null;
+          }
+          return $this->getPaidOrderDate($customer, false);
+        }
+      ),
+      new Field(
+        'woocommerce:customer:purchased-categories',
+        Field::TYPE_ENUM_ARRAY,
+        __('Purchased categories', 'mailpoet'),
+        function (CustomerPayload $payload) {
+          $customer = $payload->getCustomer();
+          return $customer ? $this->getOrderProductTermIds($customer, 'product_cat') : [];
+        },
+        [
+          'options' => $this->getTermsList('product_cat'),
+        ]
+      ),
+      new Field(
+        'woocommerce:customer:purchased-tags',
+        Field::TYPE_ENUM_ARRAY,
+        __('Purchased tags', 'mailpoet'),
+        function (CustomerPayload $payload) {
+          $customer = $payload->getCustomer();
+          return $customer ? $this->getOrderProductTermIds($customer, 'product_tag') : [];
+        },
+        [
+          'options' => $this->getTermsList('product_tag'),
+        ]
+      ),
+    ];
+  }
+
+  private function getPaidOrderDate(WC_Customer $customer, bool $fetchFirst): ?DateTimeImmutable {
+    $wpdb = $this->wordPress->getWpdb();
+    $sorting = $fetchFirst ? 'ASC' : 'DESC';
+    $statuses = array_map(function (string $status) {
+      return "wc-$status";
+    }, $this->wooCommerce->wcGetIsPaidStatuses());
+    $statusesPlaceholder = implode(',', array_fill(0, count($statuses), '%s'));
+
+    if ($this->wooCommerce->isWooCommerceCustomOrdersTableEnabled()) {
+      $statement = (string)$wpdb->prepare("
+        SELECT o.date_created_gmt
+        FROM {$wpdb->prefix}wc_orders o
+        WHERE o.customer_id = %d
+        AND o.status IN ($statusesPlaceholder)
+        AND o.total_amount > 0
+        ORDER BY o.date_created_gmt {$sorting}
+        LIMIT 1
+      ", array_merge([$customer->get_id()], $statuses));
+    } else {
+      $statement = (string)$wpdb->prepare("
+        SELECT p.post_date_gmt
+        FROM {$wpdb->prefix}posts p
+        LEFT JOIN {$wpdb->prefix}postmeta pm_total ON p.ID = pm_total.post_id AND pm_total.meta_key = '_order_total'
+        LEFT JOIN {$wpdb->prefix}postmeta pm_user ON p.ID = pm_user.post_id AND pm_user.meta_key = '_customer_user'
+        WHERE p.post_type = 'shop_order'
+        AND p.post_status IN ($statusesPlaceholder)
+        AND pm_user.meta_value = %d
+        AND pm_total.meta_value > 0
```

Comment:
Is this meta_value query really necessary? Maybe we can query also the items with `<= 0` and remove them in PHP. That should be faster imo.

Reply:
Actually, why not using
```
    wc_get_orders([
      'customer_id' => $customer->get_id(),
      'status' => $statuses,
      'limit' => 1,
      'orderby' => 'date',
      'order' => $sorting,
    ]);
```

Reply:
> Is this meta_value query really necessary? Maybe we can query also the items with <= 0 and remove them in PHP. That should be faster imo.

Fetching all orders to PHP would almost certainly be slower (fetching all order data, hydrating all of them, etc.), and could cause out-of-memory errors, so we'd need to iterate the result paginated in batches, and look for first/last order with `total > 0`. It's doable, but probably becomes more complex than the query.

> Actually, why not using

@websupporter I had that at first, but you can't do `total > 0` this way, so there really is not much of a choice.

Reply:
@websupporter I checked AW, and they seem to be doing this: https://github.com/woocommerce/automatewoo/blob/6264b07f637bd17ff44f33cba303515e8a0d1230/includes/Customer.php#L738

So, it's quite similar, although they fetch and hydrate products and then iterate those. I went a step further — directly for the data I need.